src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/EventPipeSession.cs

```diff
@@ -84,10 +95,26 @@ public async Task StopAsync(CancellationToken cancellationToken)
         private static IpcMessage CreateStartMessage(EventPipeSessionConfiguration config)
         {
             // To keep backward compatibility with older runtimes we only use newer serialization format when needed
-            // V3 has added support to disable the stacktraces
-            bool shouldUseV3 = !config.RequestStackwalk;
-            EventPipeCommandId command = shouldUseV3 ? EventPipeCommandId.CollectTracing3 : EventPipeCommandId.CollectTracing2;
-            byte[] payload = shouldUseV3 ? config.SerializeV3() : config.SerializeV2();
+            EventPipeCommandId command;
+            byte[] payload;
+            if (config.RundownKeyword != DefaultRundownKeyword && config.RundownKeyword != 0)
```

Comment:
Not requesting any change here, just a heads up.

I saw you and Justin discussing the idea that maybe the runtime would change the default rundown in the future for different versions of the runtime. If we ever did that it would break this fallback logic. Switching from V4 to V3 or V2 only works transparently because tools can predict what keywords V2/V3 rundown is going to use. Changing the rundown keywords for V2/V3 rundown would be a breaking change and I'd have no intention of doing it without a very compelling reason.

Reply:
> I saw you and Justin discussing the idea that maybe the runtime would change the default rundown in the future for different versions of the runtime. If we ever did that it would break this fallback logic.

And that's why I advocate for not checking if the specified rundown keyword was any particular value. My position is if the caller provided a rundown keyword, then use V4 unconditionally. If not provided, then use V3/V2. This is easy to explain and document.

Just because V4 is newer and will fail in the (at this moment in time) predominant runtime versions doesn't mean it should be overly optimized to avoid using V4. This isn't a critical path since callers aren't going to be trying to start event pipe sessions in tight loops or anything remotely close to that usage. Removing the extra conditions makes it clear how the callers intent is mapped to which version of the command is used.

Reply:
This PR has been dragging for weeks, I will merge it first to unblock usage/testing/integration. Feel free to continue the discussion or necessary change in another PR.

I second Noah's idea about not changing the default rundown keyword in the runtime, there is just no reason to.

For shipped runtime versions, it makes no sense to change them as a servicing change.

For newer versions, the new `CollectV4` capability allows you to specify whatever keyword necessary from the tools, so there is no need to change the default keyword so to risk a breaking change either.

I think we might want to rename the constant to legacy default keyword, to indicate that is a legacy concept.

Reply:
Not sure if the focus of my message was clear. The intended thrust was: "The rundown keywords of CollectTracing1-3 are bound by back-compat so our designs don't need to anticipate they will change".

> And that's why I advocate for not checking if the specified rundown keyword was any particular value. My position is if the caller provided a rundown keyword, then use V4 unconditionally.

I pointed to the fallback logic there only as an example of the kinds of things that break if we change the rundown keywords later. Even if we eliminate that specific check any other tool is free to make the same check in its own code. The back-compat constraints will be unchanged regardless of whether that specific check is there or not.

Reply:
> Not sure if the focus of my message was clear. The intended thrust was: "The rundown keywords of CollectTracing1-3 are bound by back-compat so our designs don't need to anticipate they will change".

That's perfectly fine. Those command versions either "collect whatever rundown the runtime happens to produce" or "do not collect rundown". That doesn't change regardless of if future runtime versions change what their definition of the default rundown keywords is.

> Even if we eliminate that specific check any other tool is free to make the same check in its own code.

And I would argue that the check is very much unnecessary if not invalid in those tools' code as well.

If a future runtime version changes its definition of what default rundown keywords are, how is dotnet/diagnostics going to adjust for it? If it's just going to remove the hard-coded default rundown keyword check, then it should have been that from the beginning (which is what I've been arguing for this entire time). The other ways I can think of are (1) doing runtime version checks (but that's not desired in NETCore.Client, according to https://github.com/dotnet/diagnostics/pull/4615#discussion_r1593238743: no policy code, just raw messaging), (2) artificially creating a V5 command just to differentiate the runtime and then pass the known default rundown keyword set for V5 and pass the original set for V4 (I would never advocate for that), or (3) have the caller prescribe different sets of rundown keywords somehow (that's going to be a mess, IMO).

I fully appreciate that this is an internal detail of the `EventPipeSession` that could be changed/fixed at any time, but I think it should be changed/fixed now in order to prevent the proliferation of what's understood to be "the" default runtime keyword set. If someone went into a specific version of the runtime and plucked out the value themselves, that was their choice to do so given they knew they are picking a runtime version specific value. If they get it from this library, they may think it's true in all cases since this library is built to be used against all runtime versions.

Thinking this through more, I think we should eliminate the notion of a default rundown keyword set from this code and have every profile prescribe its expected rundown keywords. If those happen to be what is considered the defaults from the runtime as it is today, that's fine, but I don't think the library should be prescribing what "the" default runtime keyword set is.

I think I've labored on this point with enough rationale; I will drop arguing for it after this.

Reply:
> I second Noah's idea about not changing the default rundown keyword in the runtime, there is just no reason to.

That's a policy right now. Policy makers, given enough evidence for a need to make a change, may disagree with the previous policy and change it. That's why I advocate for writing code that doesn't rely on implementation details. Additionally, it becomes a mess for documentation purposes if it does need to be changed and the library behavior changes.

> For shipped runtime versions, it makes no sense to change them as a servicing change.

No one argued for change the default rundown keywords in servicing existing runtime versions.

Reply:
> Thinking this through more, I think we should eliminate the notion of a default rundown keyword set from this code and have every profile prescribe its expected rundown keywords.

In case it isn't clear - the property `RequestRundown` (and therefore instructing the runtime to collect a default rundown) is removed. The only property we have on the profiles/source configurations is `RundownKeyword`, and there are all specified to a specific value. They just happen to all reference the internal constant `EventPipeSession.DefaultRundownKeyword` just for de-duplication purpose. 

It is exactly the idea of "every profile prescribes its expected rundown keywords."

Reply:
> That's why I advocate for writing code that doesn't rely on implementation details.

I am trying to clarify I do not consider this to be an implementation detail. Previously it wasn't explicitly written in the spec and I have fixed that:
https://github.com/dotnet/diagnostics/pull/4658